File: Web/Web/Script/Handler/MessageHandler/DatabaseMessage.js
//Copyright(c) 2015,HIT All rights reserved.
//Des:DatabaseMessage Handler for Local database
//Author:Irlovan   
//Date:2015-03-26

Irlovan.Include.Using("Script/Message/Message.js");

/**Construction**/
Irlovan.Handler.DatabaseMessage = function (session) {
    this.RootTag = "Database";
    this.ItemTag = "Item";
    this.Session = session;
    this.Callback = null;
}

/**Handle message from server**/
Irlovan.Handler.DatabaseMessage.prototype.Handle = function (xDocument) {
    var element = xDocument.documentElement;
    if (element.nodeName != this.RootTag) { return; }
    this.HandleDatabase(element);
    if (this.Callback) { this.Callback(); }
}

/**Subcribe for Local database**/
Irlovan.Handler.DatabaseMessage.prototype.Subcribe = function (callback) {
    this.Callback = callback;
    var doc = Irlovan.Lib.XML.CreateXDocument(this.RootTag);
    this.Session.Send(Irlovan.Lib.XML.ToString(doc));
}

/**Handle Database Init**/
Irlovan.Handler.DatabaseMessage.prototype.HandleDatabase = function (element) {
    var dataItem = element.getElementsByTagName(this.ItemTag);
    if ((dataItem == null) || (dataItem.length == 0)) { return; }
    var result = [];
    for (var i = 0; i < dataItem.length; i++) {
        result.push(Irlovan.Lib.XML.ToString(dataItem[i]));
    }
    Irlovan.Global.RealtimeDataList = result;
    Irlovan.Global.LoadBar.Off();
}

/**Dispose**/
Irlovan.Handler.DatabaseMessage.prototype.Dispose = function () { }

File: MessageHandler/Type/DatabaseMessage.cs
///Copyright(c) 2015,Irlovan All rights reserved.
///Summary:
///Author:Irlovan
///Date:2015-11-12
///Description:
///Modification:      

using Irlovan.Canal;
using Irlovan.Message;
using System.Xml.Linq;

namespace Irlovan.Handlers
{
    internal class DatabaseMessage : Handler
    {

        #region Structure

        /// <summary>
        /// Handler for realtime Event
        /// </summary>
        internal DatabaseMessage(IServerSession session, LocalInterface.LocalInterface local)
            : base(session, local) {
        }

        #endregion Structure

        #region Field

        internal const string Name = "Database";
        internal const string DatabaseItemPara = "Item";

        #endregion Field

        #region Function

        /// <summary>
        /// Handler message
        /// </summary>
        /// <param name="session"></param>
        /// <param name="message"></param>
        /// <returns></returns>
        public override bool Handle(IServerSession session, XElement message) {
            if (!base.Handle(session, message)) { return false; };
            Session.Send(GetAllData());
            return true;
        }

        /// <summary>
        /// Get All Data
        /// </summary>
        /// <returns></returns>
        private string GetAllData() {
            XElement result = new XElement(Name);
            foreach (var item in LocalInterface.Source.AcquireAll(true)) {
                string value = (item.Value == null) ? string.Empty : item.Value.ToString();
                XElement dataMessage = (new IndustryDataMessage(item.FullName, value, item.DataType, item.TimeStamp, item.Description, item.Quality)).ToXML(FormatEnum.Typic);
                dataMessage.Name = DatabaseItemPara;
                result.Add(dataMessage);
            }
            return result.ToString();
        }

        /// <summary>
        /// Dispose
        /// </summary>
        public override void Dispose() {
            base.Dispose();
        }

        #endregion Function

    }
}